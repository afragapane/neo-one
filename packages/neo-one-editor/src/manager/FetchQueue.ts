import fetch from 'cross-fetch';

export interface QueueData<T> {
  readonly url: string;
  readonly handleResponse: (response: Response) => Promise<T>;
  readonly resolve: (result: T) => void;
  readonly reject: (error: Error) => void;
}

interface FetchQueueOptions {
  readonly fetchConcurrency: number;
}

export class FetchQueue<T> {
  private readonly fetchConcurrency: number;
  private readonly mutableQueue: Array<QueueData<T>>;
  private mutableRunning: number;

  public constructor(options?: FetchQueueOptions) {
    this.fetchConcurrency = options === undefined ? 10 : options.fetchConcurrency;
    this.mutableQueue = [];
    this.mutableRunning = 0;
  }

  public push(queueData: QueueData<T>) {
    this.mutableQueue.push(queueData);
  }

  public advance() {
    if (this.mutableRunning > this.fetchConcurrency) {
      return;
    }
    this.testSpy(this.mutableRunning);

    const entry = this.mutableQueue.shift();
    if (entry !== undefined) {
      this.mutableRunning += 1;
      fetch(entry.url)
        .then(async (res) => {
          const result = await entry.handleResponse(res);
          entry.resolve(result);
          this.mutableRunning -= 1;
          this.advance();
        })
        .catch(entry.reject);
    }
  }

  public testSpy(_testMutableRemaining: number) {
    // do nothing;
  }
}
