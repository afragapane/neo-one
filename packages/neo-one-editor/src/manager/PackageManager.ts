import { FileSystem } from '@neo-one/local-browser';
import fetch from 'cross-fetch';
import * as path from 'path';
import { Observable, Subscription } from 'rxjs';
import { distinctUntilChanged, map, switchMap } from 'rxjs/operators';
import { Dependencies, DependencyInfo, PackageJSON, ResolvedDependencies, Resolver } from './Resolver';

const PACKAGE_JSON = 'package.json';

interface FetchPackageInfo {
  readonly name: string;
  readonly version: string;
  readonly file?: string;
}

interface NMPackageInfo {
  readonly depInfo: DependencyInfo;
  readonly pkgPath: string;
}

interface Module {
  readonly name: string;
  readonly fileName: string;
  readonly version: string;
  readonly dependency: string;
  readonly packageJSON: string;
  readonly types?: string;
  readonly typeFile: string;
}

export class PackageManager {
  public readonly dependencies$: Observable<ResolvedDependencies>;
  public readonly packages$: Observable<{ readonly [name: string]: Module }>;
  private readonly fs: FileSystem;
  private readonly subscription: Subscription;

  public constructor({
    fs,
    packageJSON$,
  }: {
    readonly packageJSON$: Observable<PackageJSON>;
    readonly fs: FileSystem;
  }) {
    this.fs = fs;
    this.dependencies$ = packageJSON$.pipe(
      distinctUntilChanged(),
      switchMap(async (packageJSON) =>
        this.resolveDependencies(packageJSON.dependencies === undefined ? {} : packageJSON.dependencies),
      ),
    );
    this.packages$ = this.dependencies$.pipe(switchMap(async (deps) => this.fetchPackages(deps, 'node_modules')));
    this.subscription = this.packages$
      .pipe(map(async (packages) => this.writeModules({ packages, pkgPath: 'node_modules' })))
      .subscribe();
  }

  public cleanup() {
    this.subscription.unsubscribe();
  }

  public async resolveDependencies(dependencies: Dependencies): Promise<ResolvedDependencies> {
    const resolver = new Resolver();

    return resolver.resolve(dependencies);
  }

  private async fetchPackages(deps: ResolvedDependencies, pkgPath: string) {
    const packages = await Promise.all(
      Object.values(deps).map(async (depInfo) => this.fetchPackage({ depInfo, pkgPath })),
    );

    return packages.reduce<{ readonly [name: string]: Module }>((acc, pkg) => ({ ...acc, [pkg.name]: pkg }), {});
  }

  private async fetchPackage({ depInfo, pkgPath }: NMPackageInfo): Promise<Module> {
    let file;
    if (depInfo.main !== undefined && depInfo.main.includes('.')) {
      file = depInfo.main;
    }
    const [pkg, pkgJSON, types] = await Promise.all([
      this.downloadDependency({ name: depInfo.name, version: depInfo.version, file }),
      this.downloadDependency({ name: depInfo.name, version: depInfo.version, file: PACKAGE_JSON }),
      depInfo.types === undefined
        ? Promise.resolve(undefined)
        : this.downloadDependency({ name: depInfo.name, version: depInfo.version, file: depInfo.types }),
    ]);

    if (Object.keys(depInfo.dependencies).length !== 0) {
      const subPkgs = await this.fetchPackages(depInfo.dependencies, path.join(pkgPath, name));
      await this.writeModules({ packages: subPkgs, pkgPath: path.resolve('node_modules', name, 'node_modules') });
    }

    return {
      name: depInfo.name,
      fileName: file === undefined ? 'index.js' : file,
      version: depInfo.version,
      dependency: pkg,
      packageJSON: pkgJSON,
      types,
      typeFile: depInfo.types === undefined ? 'index.d.ts' : depInfo.types,
    };
  }

  private getNPMUrl({ name, version, file }: FetchPackageInfo): string {
    const urlBase = `https://cdn.jsdelivr.net/npm/${name}@${version}`;

    return file === undefined ? urlBase : `${urlBase}/${file}`;
  }

  private async downloadDependency({ name, version, file }: FetchPackageInfo) {
    try {
      const pkg = await fetch(this.getNPMUrl({ name, version, file }));

      return pkg.text();
      // tslint:disable-next-line no-unused
    } catch (error) {
      throw new Error(`Could not find module ${name}@${version}`);
    }
  }

  private async writeModule({
    name,
    pkg,
    pkgPath,
  }: {
    readonly name: string;
    readonly pkg: Module;
    readonly pkgPath: string;
  }) {
    await this.fs.writeFile(path.resolve(pkgPath, name, pkg.fileName), pkg.dependency);
    await this.fs.writeFile(path.resolve(pkgPath, name, 'package.json'), pkg.packageJSON);
    if (pkg.types !== undefined) {
      await this.fs.writeFile(path.resolve(pkgPath, name, pkg.typeFile), pkg.types);
    }
  }

  private async writeModules({
    packages,
    pkgPath,
  }: {
    readonly packages: { readonly [name: string]: Module };
    readonly pkgPath: string;
  }) {
    Object.entries(packages).forEach(async ([name, pkg]) => this.writeModule({ name, pkg, pkgPath }));
  }
}
