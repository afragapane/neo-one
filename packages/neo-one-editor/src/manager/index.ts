export * from './PackageManager';
export * from './Resolver';
