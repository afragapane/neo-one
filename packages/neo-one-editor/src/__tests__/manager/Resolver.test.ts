/* @jest-environment jsdom */
import { Resolver } from '../../manager';

const dependencies = {
  'bignumber.js': '^7.2.1',
  react: '^16.5.1',
  reakit: '0.15.7',
  rxjs: '^6.3.3',
  'react-dom': '^16.5.1',
  'styled-components': '^3.4.10',
};

describe('resolver', () => {
  test('resolve deps', async () => {
    const resolver = new Resolver();

    const resolvedDependencies = await resolver.resolve(dependencies);
    expect(resolvedDependencies).toMatchSnapshot();
  });
});
