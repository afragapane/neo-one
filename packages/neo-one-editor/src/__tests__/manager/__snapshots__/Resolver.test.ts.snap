// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolver resolve deps 1`] = `
Object {
  "@emotion/is-prop-valid": Object {
    "dependencies": Object {},
    "main": "dist/is-prop-valid.cjs.js",
    "name": "@emotion/is-prop-valid",
    "types": undefined,
    "version": "0.6.8",
  },
  "@emotion/memoize": Object {
    "dependencies": Object {},
    "main": "dist/memoize.cjs.js",
    "name": "@emotion/memoize",
    "types": undefined,
    "version": "0.6.6",
  },
  "asap": Object {
    "dependencies": Object {},
    "main": "./asap.js",
    "name": "asap",
    "types": undefined,
    "version": "2.0.6",
  },
  "base64-js": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "base64-js",
    "types": undefined,
    "version": "1.3.0",
  },
  "bignumber.js": Object {
    "dependencies": Object {},
    "main": "bignumber",
    "name": "bignumber.js",
    "types": "bignumber.d.ts",
    "version": "7.2.1",
  },
  "buffer": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "buffer",
    "types": "index.d.ts",
    "version": "5.2.1",
  },
  "constate": Object {
    "dependencies": Object {},
    "main": "dist/constate.cjs.js",
    "name": "constate",
    "types": "dist/ts/src",
    "version": "0.9.0",
  },
  "core-js": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "core-js",
    "types": undefined,
    "version": "1.2.7",
  },
  "css-color-keywords": Object {
    "dependencies": Object {},
    "main": undefined,
    "name": "css-color-keywords",
    "types": undefined,
    "version": "1.0.0",
  },
  "css-to-react-native": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "css-to-react-native",
    "types": undefined,
    "version": "2.2.2",
  },
  "encoding": Object {
    "dependencies": Object {},
    "main": "lib/encoding.js",
    "name": "encoding",
    "types": undefined,
    "version": "0.1.12",
  },
  "fbjs": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "fbjs",
    "types": undefined,
    "version": "0.8.17",
  },
  "has-flag": Object {
    "dependencies": Object {},
    "main": undefined,
    "name": "has-flag",
    "types": undefined,
    "version": "1.0.0",
  },
  "hoist-non-react-statics": Object {
    "dependencies": Object {},
    "main": "dist/hoist-non-react-statics.cjs.js",
    "name": "hoist-non-react-statics",
    "types": "index.d.ts",
    "version": "2.5.5",
  },
  "iconv-lite": Object {
    "dependencies": Object {},
    "main": "./lib/index.js",
    "name": "iconv-lite",
    "types": "./lib/index.d.ts",
    "version": "0.4.24",
  },
  "ieee754": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "ieee754",
    "types": undefined,
    "version": "1.1.12",
  },
  "is-stream": Object {
    "dependencies": Object {},
    "main": undefined,
    "name": "is-stream",
    "types": undefined,
    "version": "1.1.0",
  },
  "isomorphic-fetch": Object {
    "dependencies": Object {},
    "main": "fetch-npm-node.js",
    "name": "isomorphic-fetch",
    "types": undefined,
    "version": "2.2.1",
  },
  "js-tokens": Object {
    "dependencies": Object {},
    "main": undefined,
    "name": "js-tokens",
    "types": undefined,
    "version": "4.0.0",
  },
  "loose-envify": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "loose-envify",
    "types": undefined,
    "version": "1.4.0",
  },
  "node-fetch": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "node-fetch",
    "types": undefined,
    "version": "1.7.3",
  },
  "object-assign": Object {
    "dependencies": Object {},
    "main": undefined,
    "name": "object-assign",
    "types": undefined,
    "version": "4.1.1",
  },
  "popper.js": Object {
    "dependencies": Object {},
    "main": "dist/umd/popper.js",
    "name": "popper.js",
    "types": "index.d.ts",
    "version": "1.14.4",
  },
  "postcss-value-parser": Object {
    "dependencies": Object {},
    "main": "lib/index.js",
    "name": "postcss-value-parser",
    "types": undefined,
    "version": "3.3.1",
  },
  "promise": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "promise",
    "types": undefined,
    "version": "7.3.1",
  },
  "prop-types": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "prop-types",
    "types": undefined,
    "version": "15.6.2",
  },
  "react": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "react",
    "types": undefined,
    "version": "16.5.2",
  },
  "react-dom": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "react-dom",
    "types": undefined,
    "version": "16.5.2",
  },
  "react-is": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "react-is",
    "types": undefined,
    "version": "16.5.2",
  },
  "reakit": Object {
    "dependencies": Object {},
    "main": "lib/index.js",
    "name": "reakit",
    "types": "ts",
    "version": "0.15.7",
  },
  "rxjs": Object {
    "dependencies": Object {},
    "main": "./index.js",
    "name": "rxjs",
    "types": "./index.d.ts",
    "version": "6.3.3",
  },
  "safer-buffer": Object {
    "dependencies": Object {},
    "main": "safer.js",
    "name": "safer-buffer",
    "types": undefined,
    "version": "2.1.2",
  },
  "schedule": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "schedule",
    "types": undefined,
    "version": "0.5.0",
  },
  "setimmediate": Object {
    "dependencies": Object {},
    "main": "setImmediate.js",
    "name": "setimmediate",
    "types": undefined,
    "version": "1.0.5",
  },
  "styled-components": Object {
    "dependencies": Object {},
    "main": "dist/styled-components.cjs.js",
    "name": "styled-components",
    "types": "typings/styled-components.d.ts",
    "version": "3.4.10",
  },
  "styled-tools": Object {
    "dependencies": Object {},
    "main": "dist/cjs/index.js",
    "name": "styled-tools",
    "types": "types/index.d.ts",
    "version": "1.5.2",
  },
  "stylis": Object {
    "dependencies": Object {},
    "main": "stylis.js",
    "name": "stylis",
    "types": "stylis.d.ts",
    "version": "3.5.3",
  },
  "stylis-rule-sheet": Object {
    "dependencies": Object {},
    "main": "index.js",
    "name": "stylis-rule-sheet",
    "types": undefined,
    "version": "0.0.10",
  },
  "supports-color": Object {
    "dependencies": Object {},
    "main": undefined,
    "name": "supports-color",
    "types": undefined,
    "version": "3.2.3",
  },
  "tslib": Object {
    "dependencies": Object {},
    "main": "tslib.js",
    "name": "tslib",
    "types": "tslib.d.ts",
    "version": "1.9.3",
  },
  "ua-parser-js": Object {
    "dependencies": Object {},
    "main": "src/ua-parser.js",
    "name": "ua-parser-js",
    "types": undefined,
    "version": "0.7.18",
  },
  "whatwg-fetch": Object {
    "dependencies": Object {},
    "main": "./dist/fetch.umd.js",
    "name": "whatwg-fetch",
    "types": undefined,
    "version": "3.0.0",
  },
}
`;
